internal/app/kconfig-controller/controller/knativeservicebinding/knativeservicebinding.go

```diff
@@ -0,0 +1,286 @@
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Get the KnativeServiceBinding resource with this namespace/name
+	knativeServiceBinding, err := c.knativeServiceBindingLister.KnativeServiceBindings(namespace).Get(name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			runtime.HandleError(fmt.Errorf("knativeServiceBinding '%s' in work queue no longer exists", key))
+			return nil
+		}
+		return err
+	}
+
+	err = c.processKnativeServiceBinding(knativeServiceBinding)
+	if err != nil {
+		return err
+	}
+
+	c.recorder.Event(knativeServiceBinding, corev1.EventTypeNormal, controller.SuccessSynced, controller.MessageDeploymentResourceSynced)
+	return nil
+}
+
+func (c *Controller) processKnativeServiceBinding(knativeServiceBinding *kcv1alpha1.KnativeServiceBinding) error {
+	// convert kconfigEnvs map to array of values
+	kconfigEnvs := []kcv1alpha1.KconfigEnvs{}
+	for _, kconfigEnv := range knativeServiceBinding.Spec.KconfigEnvsMap {
+		kconfigEnvs = append(kconfigEnvs, kconfigEnv)
+	}
+
+	// Sort env array
+	sort.Sort(util.ByLevel(kconfigEnvs))
+
+	// Create final env array
+	envArray := []corev1.EnvVar{}
+	// envRefsVersions tracks changes to refs among various KconfigEnvs for updating deployment
+	// template when reference value changes but no keys have changed. It is composed of each
+	// KconfigEnv's envRefsVersion concatenated into a single string
+	envRefsVersions := ""
+	for _, kconfigEnv := range kconfigEnvs {
+		envRefsVersions += strconv.FormatInt(kconfigEnv.EnvRefsVersion, 10)
+		envArray = append(envArray, kconfigEnv.Envs...)
+	}
+
+	knativeService, err := c.knativeServicelister.Services(knativeServiceBinding.Namespace).Get(knativeServiceBinding.Name)
+	if err != nil {
+		return err
+	}
+	if knativeService.GetAnnotations()[controller.KconfigEnabledDeploymentAnnotation] != "true" {
+		klog.Infof("knativeService %s %s does not have kconfig enabled", knativeService.GetNamespace(), knativeService.GetName())
+		return nil
+	}
+	knativeServiceCopy := knativeService.DeepCopy()
+
+	if knativeServiceCopy.Spec.RunLatest != nil {
+		knativeServiceCopy.Spec.RunLatest.Configuration = applyKconfigEnvToConfiguration(knativeServiceCopy.Spec.RunLatest.Configuration, envArray, envRefsVersions)
+	}
+	if knativeServiceCopy.Spec.Release != nil {
+		knativeServiceCopy.Spec.Release.Configuration = applyKconfigEnvToConfiguration(knativeServiceCopy.Spec.Release.Configuration, envArray, envRefsVersions)
+	}
+
+	if _, err = c.knclient.Serving().Services(knativeServiceCopy.Namespace).Update(knativeServiceCopy); err != nil {
+		return err
+	}
+	return nil
+}
+
+func applyKconfigEnvToConfiguration(configuration knv1alpha1.ConfigurationSpec, envArray []corev1.EnvVar, envRefsVersions string) knv1alpha1.ConfigurationSpec {
+	// I dont get why this is an error when RevisionTemplate is defined as a pointer to a RevisionTemplateSpec (should be able to be nil)
```

Comment:
Is this a bug in the client or something?

Reply:
Good find. I forgot I left this comment in here. It turns out, I was going by the current documentation on where revisionTemplate is a pointer but that was a recent change that was after the 0.5.0 version of the client I'm using. I'll remove those comment lines.